File: backend/src/main/java/pers/eastwind/billmanager/attach/service/AttachMapService.java
package pers.eastwind.billmanager.attach.service;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import pers.eastwind.billmanager.attach.model.AttachmentDTO;
import pers.eastwind.billmanager.attach.model.AttachmentType;
import pers.eastwind.billmanager.attach.util.FileUtil;
import pers.eastwind.billmanager.attach.util.OfficeFileUtil;
import pers.eastwind.billmanager.common.exception.BizException;

import java.nio.file.Path;
import java.util.List;

/**
 * 附件映射服务
 */
@Slf4j
@Service
public class AttachMapService {
    private final OCRService ocrService;
    private final List<AttachMapRule<?>> mapRules;
    private final AttachmentService attachmentService;

    public AttachMapService(OCRService ocrService, List<AttachMapRule<?>> mapRules, AttachmentService attachmentService) {
        this.ocrService = ocrService;
        this.mapRules = mapRules;
        this.attachmentService = attachmentService;
    }

    /**
     * 附件映射为对象
     * @param attachment 附件
     * @param clazz 目标对象类
     * @return 对象
     */
    public <T> T map(AttachmentDTO attachment, Class<T> clazz) {
        Path path = attachmentService.getAbsolutePath(Path.of(attachment.getRelativePath()), attachment.isTemp());
        switch (attachment.getType()) {
            case IMAGE, PDF -> {
                if (attachment.getType() == AttachmentType.PDF) {
                    Path target = attachmentService.createTempFile("", ".jpg");
                    FileUtil.convertPDFToImage(path, target);
                    path = target;
                }
                List<String> texts = ocrService.parseImage(path);
                for (AttachMapRule<?> mapRule : mapRules) {
                    var cur = mapRule.mapFromOCR(texts);
                    if (cur != null) {
                        try {
                            return clazz.cast(cur);
                        } catch (ClassCastException _) {
                            log.error("{} OCR 类型转换失败", mapRule);
                        }
                    }
                }
                throw new BizException("未配置映射规则");
            }
            case EXCEL -> {
                List<List<String>> rows = OfficeFileUtil.parseExcel(path);
                for (AttachMapRule<?> mapRule : mapRules) {
                    var cur = mapRule.mapFromExcel(rows);
                    if (cur != null) {
                        try {
                            return clazz.cast(cur);
                        } catch (ClassCastException _) {
                            log.error("{} Excel 类型转换失败", mapRule);
                        }
                    }
                }
                throw new BizException("未配置映射规则");
            }
            default -> throw new BizException("不支持该附件类型映射");
        }
    }
}

File: backend-go/internal/attach/service/attach_map_service.go
package service

import (
	"backend-go/internal/attach/files"
	"backend-go/internal/attach/model"
	"backend-go/internal/common/cache"
	"backend-go/internal/common/errs"
	"sync"
)

// AttachMapRule 附件映射规则
type AttachMapRule interface {
	// MapFromOCR 从 OCR 识别结果映射到实体, 返回 nil 表示无法映射
	MapFromOCR(texts *[]string) (any, error)
	// MapFromExcel 从 Excel 解析结果映射到实体, 返回 nil 表示无法映射
	MapFromExcel(rows *[][]string) (any, error)
}

// AttachMapService 映射服务
type AttachMapService struct {
	cache         cache.Cache
	ocrService    *OCRService
	attachService *AttachmentService
	rules         []AttachMapRule
	lock          sync.RWMutex
}

func NewAttachMapService(cache cache.Cache, ocrService *OCRService, attachService *AttachmentService) *AttachMapService {
	return &AttachMapService{
		cache:         cache,
		ocrService:    ocrService,
		attachService: attachService,
		rules:         make([]AttachMapRule, 0),
		lock:          sync.RWMutex{},
	}
}

// RegisterRule 注册映射规则实例
func (s *AttachMapService) RegisterRule(rule AttachMapRule) {
	s.lock.Lock()
	defer s.lock.Unlock()
	s.rules = append(s.rules, rule)
}

func (s *AttachMapService) MapTo(attach *model.AttachmentDTO) (any, error) {
	if attach == nil {
		return nil, errs.NewBizError("附件为空")
	}
	path, err := s.attachService.GetAbsolutePath(attach.RelativePath, attach.Temp)
	if err != nil {
		return nil, err
	}
	switch attach.Type {
	case model.AttachTypeImage, model.AttachTypePDF:
		if attach.Type == model.AttachTypePDF {
			target, err := s.attachService.CreateTempFile(s.cache, "", ".jpg")
			if err != nil {
				return nil, err
			}
			if err = files.ConvertPDFToImage(path, target); err != nil {
				return nil, err
			}
			path = target
		}
		texts, err := s.ocrService.ParseImage(path)
		if err != nil {
			return nil, err
		}
		s.lock.RLock()
		defer s.lock.RUnlock()
		for _, rule := range s.rules {
			if cur, err := rule.MapFromOCR(texts); err == nil && cur != nil {
				return cur, nil
			}
		}
		return nil, errs.NewBizError("未配置映射规则")
	case model.AttachTypeExcel:
		rows, err := files.ParseExcel(path)
		if err != nil {
			return nil, err
		}
		s.lock.RLock()
		defer s.lock.RUnlock()
		for _, rule := range s.rules {
			if cur, err := rule.MapFromExcel(rows); err == nil && cur != nil {
				return cur, nil
			}
		}
		return nil, errs.NewBizError("未配置映射规则")
	default:
		return nil, errs.NewBizError("不支持的映射文件类型")
	}
}
